File: python/happy_number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        number_set: set[int] = set()
        while n not in number_set:
            if n == 1:
                return True
            number_set.add(n)
            n = self.sumOfSquares(n)

        return False

    def sumOfSquares(self, n: int) -> int:
        output: int = 0
        while n > 0:
            digit: int = n % 10
            output += digit**2
            n = n // 10

        return output


def test_isHappy_case_1():
    # Arrange
    input = 19
    expected_output = True

    # Act
    solution = Solution()
    actual_output = solution.isHappy(input)

    # Assert
    assert actual_output == expected_output


def test_isHappy_case_2():
    # Arrange
    input = 2
    expected_output = False

    # Act
    solution = Solution()
    actual_output = solution.isHappy(input)

    # Assert
    assert actual_output == expected_output


def test_isHappy_case_3():
    # Arrange
    input = 7
    expected_output = True

    # Act
    solution = Solution()
    actual_output = solution.isHappy(input)

    # Assert
    assert actual_output == expected_output

File: rust/src/code/happy_number.rs
use std::collections::HashSet;

struct Solution {}

impl Solution {
    pub fn is_happy(mut n: i32) -> bool {
        let mut set: HashSet<i32> = HashSet::new();
        while !set.contains(&n) {
            if n == 1 {
                return true;
            }

            set.insert(n);
            n = Self::sum_of_squares(n);
        }

        false
    }

    fn sum_of_squares(mut n: i32) -> i32 {
        let mut output = 0;
        while n > 0 {
            let digit = n % 10;
            output += digit.pow(2);
            n = n / 10;
        }

        output
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_is_happy_case_1() {
        // arrange
        let n = 19;
        let expected = true;

        // act
        let actual = Solution::is_happy(n);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_is_happy_case_2() {
        // arrange
        let n = 2;
        let expected = false;

        // act
        let actual = Solution::is_happy(n);

        // assert
        assert_eq!(actual, expected);
    }
}
